Source/Core/Common/HandShake.h

```diff
@@ -0,0 +1,84 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <atomic>
+#include <mutex>
+#include <optional>
+
+template <typename Inner>
+struct HandShake
+{
+public:
+  class ReaderGuard
+  {
+  private:
+    HandShake& parent;
+    std::unique_lock<std::mutex> guard;
+
+  public:
+    // a concurrency guru should probably check whether these memory orders are correct
```

Comment:
You should probably know how to properly do this sort of thing if you're dealing with threading at all, just an FYI. This is, to me, a tremendous red flag for what's to come.

Reply:
I *have* thought about all ways (that I can think of) how this interaction can play out for a few hours, I just haven't gotten to having two threads use this yet (real-world systems might not behave like my understanding of the memory model, after all). I hope not to need more than this primitive and a handful of mutexes anyway, so there is not really 'more to come' except considerations for I$ sync and deadlocking.